src/ModelContextProtocol.Analyzers/XmlToDescriptionGenerator.cs

```diff
@@ -0,0 +1,386 @@
+            context.AddSource(GeneratedFileName, SourceText.From(source, Encoding.UTF8));
+        }
+    }
+
+    private static XmlDocumentation? ExtractXmlDocumentation(IMethodSymbol methodSymbol, SourceProductionContext context)
+    {
+        string? xmlDoc = methodSymbol.GetDocumentationCommentXml();
+        if (string.IsNullOrWhiteSpace(xmlDoc))
+        {
+            return null;
+        }
+
+        try
+        {
+            if (XDocument.Parse(xmlDoc).Element("member") is not { } memberElement)
+            {
+                return null;
+            }
+
+            var summary = CleanXmlDocText(memberElement.Element("summary")?.Value);
+            var remarks = CleanXmlDocText(memberElement.Element("remarks")?.Value);
+            var returns = CleanXmlDocText(memberElement.Element("returns")?.Value);
+
+            // Combine summary and remarks for method description.
+            var methodDescription = 
+                string.IsNullOrWhiteSpace(remarks) ? summary :
+                string.IsNullOrWhiteSpace(summary) ? remarks :
+                $"{summary}\n{remarks}";
+
+            Dictionary<string, string> paramDocs = new(StringComparer.Ordinal);
+            foreach (var paramElement in memberElement.Elements("param"))
+            {
+                var name = paramElement.Attribute("name")?.Value;
+                var value = CleanXmlDocText(paramElement.Value);
+                if (!string.IsNullOrWhiteSpace(name) && !string.IsNullOrWhiteSpace(value))
+                {
+                    paramDocs[name!] = value;
+                }
+            }
+
+            // Return documentation even if empty - we'll still generate the partial implementation
+            return new XmlDocumentation(methodDescription ?? string.Empty, returns ?? string.Empty, paramDocs);
+        }
+        catch (System.Xml.XmlException)
+        {
+            // Emit warning for invalid XML
+            context.ReportDiagnostic(Diagnostic.Create(
+                Diagnostics.InvalidXmlDocumentation,
+                methodSymbol.Locations.FirstOrDefault(),
+                methodSymbol.Name));
+            return null;
+        }
+    }
+
+    private static string CleanXmlDocText(string? text)
+    {
+        if (string.IsNullOrWhiteSpace(text))
+        {
+            return string.Empty;
+        }
+
+        // Remove leading/trailing whitespace and normalize line breaks
+        var lines = text!.Split('\n')
+            .Select(line => line.Trim())
+            .Where(line => !string.IsNullOrEmpty(line));
+
+        return string.Join(" ", lines).Trim();
+    }
+
+    private static string GenerateSourceFile(
+        Compilation compilation,
+        List<(IMethodSymbol MethodSymbol, MethodDeclarationSyntax MethodDeclaration, XmlDocumentation? XmlDocs)> methods,
+        INamedTypeSymbol descriptionAttribute)
+    {
+        var sb = new StringBuilder();
```

Comment:
Common practice dictates that we separate out model extractions and source code formatting into separate components/files using the "parser" vs. "emitter" naming convention.

Reply:
Given how small this is, I don't see value in splitting it across multiple files.